### ISSUE TITLE
Connection reset when making a request to an upstream

### ISSUE THREAD

**Issue by mariusehr1** [hashicorp/consul#9276]
#### Overview of the Issue

My setup is three datacenters: a primary DC hosting Consul and Vault

Vault is reachable by any server within the mesh via a public IP.

DC2 is composed of VMs : 1 mesh gateway 1 server 1 agent 1 service running : payments with a sidecar

DC3 is a k8s DC , consul is deployed with Helm. I have a mesh gateway, servers, sync catalog, clients and two services : Grafana and Nginx

Payments has an upstream to grafana and nginx but it does not work : I get the following : 
```
user@payments:/$ curl localhost:9091 <= grafana upstream but nginx behaves the same
curl: (56) Recv failure: Connection reset by peer
```

Even though all the federation is working : UI says that everything is functionnal and every component registers properly

I am using Vault as a CA provider, and each secondary and primary DCs are properly setup to use a different `connect-intermediate` (e.g dc2/connect-intermediate etc..)

ACL are activated as well as auto_encrypt everywhere, intention are enabled as well but allowing any service to any service for debugging purposes.

#### Reproduction Steps

Steps to reproduce this issue, eg:
1. Register a service with an upstream as follows, it also registers a service-default as well as a service-resolver to make it reachable from any DC : 
```
service {
  name = "payments"
  id = "payments"
  address = "IP"
  port = 9090
  connect {
    sidecar_service {
      port = 20000
      check {
        name = "Connect Envoy Sidecar"
        tcp = "IP:20000"
        interval ="10s"
      }
      proxy {
        upstreams {
          destination_name = "grafana"
	  datacenter = "dc3"
          local_bind_address = "127.0.0.1"
          local_bind_port = 9091
          }
        }
      }
    }
  }
```
1. Setting the upstream from payments to an other service
1. Run `curl localhost:9091`
1. Getting  `curl: (56) Recv failure: Connection reset by peer`
Its worth noting that I am able to reach services from any DC as long as its in the same DC. I am registering service-resolver for each services as well.

### Consul info for both Client and Server

<details>
  <summary>Client info</summary>

```
agent:
	check_monitors = 0
	check_ttls = 0
	checks = 0
	services = 0
build:
	prerelease =
	revision = 1e03567d
	version = 1.8.5
consul:
	acl = enabled
	bootstrap = false
	known_datacenters = 4
	leader = true
	leader_addr = 192.168.11.126:8300
	server = true
raft:
	applied_index = 8173
	commit_index = 8173
	fsm_pending = 0
	last_contact = 0
	last_log_index = 8173
	last_log_term = 2
	last_snapshot_index = 0
	last_snapshot_term = 0
	latest_configuration = [{Suffrage:Voter ID:95f81bf7-8d9e-fc7e-ddbb-9f405c67e882 Address:192.168.53.117:8300} {Suffrage:Voter ID:0da2a6ab-e61a-c059-e343-b8af785827a0 Address:192.168.11.126:8300}]
	latest_configuration_index = 0
	num_peers = 1
	protocol_version = 3
	protocol_version_max = 3
	protocol_version_min = 0
	snapshot_version_max = 1
	snapshot_version_min = 0
	state = Leader
	term = 2
runtime:
	arch = amd64
	cpu_count = 2
	goroutines = 239
	max_procs = 2
	os = linux
	version = go1.14.9
serf_lan:
	coordinate_resets = 0
	encrypted = true
	event_queue = 0
	event_time = 2
	failed = 0
	health_score = 0
	intent_queue = 0
	left = 0
	member_time = 5
	members = 4
	query_queue = 0
	query_time = 1
serf_wan:
	coordinate_resets = 0
	encrypted = true
	event_queue = 0
	event_time = 1
	failed = 2
	health_score = 0
	intent_queue = 0
	left = 0
	member_time = 531
	members = 8
	query_queue = 0
	query_time = 1
```

</details>

<details>
  <summary>Server info</summary>

```
agent:
	check_monitors = 0
	check_ttls = 0
	checks = 5
	services = 5
build:
	prerelease =
	revision = 1e03567d
	version = 1.8.5
consul:
	acl = enabled
	known_servers = 2
	server = false
runtime:
	arch = amd64
	cpu_count = 2
	goroutines = 229
	max_procs = 2
	os = linux
	version = go1.14.9
serf_lan:
	coordinate_resets = 0
	encrypted = true
	event_queue = 0
	event_time = 2
	failed = 0
	health_score = 0
	intent_queue = 0
	left = 0
	member_time = 5
	members = 4
	query_queue = 0
	query_time = 1
```

</details>

### Operating system and Environment details

Running on Ubuntu 18.04 on the VMs 

running on the default EKS image on k8s

All the network connectivity is functionnal, each mesh-gateway can talk to any other one.

### Log Fragments
Running envoy in debug mod I get the following when trying to call the upstream: 

<details>
<summary>Envoy logs : </summary>

```
Nov 25 15:42:45 ip-172-16-64-11 consul[198775]: [2020-11-25 15:42:45.299][198794][debug][filter] [external/envoy/source/common/tcp_proxy/tcp_proxy.cc:251] [C17] new tcp proxy session
Nov 25 15:42:45 ip-172-16-64-11 consul[198775]: [2020-11-25 15:42:45.299][198794][debug][filter] [external/envoy/source/common/tcp_proxy/tcp_proxy.cc:395] [C17] Creating connection to cluster grafana.default.dc3.internal.e5b63d86-f01b-3883-0b42-7ef5dd8700f9.consul
Nov 25 15:42:45 ip-172-16-64-11 consul[198775]: [2020-11-25 15:42:45.299][198794][debug][pool] [external/envoy/source/common/tcp/conn_pool.cc:83] creating a new connection
Nov 25 15:42:45 ip-172-16-64-11 consul[198775]: [2020-11-25 15:42:45.300][198794][debug][pool] [external/envoy/source/common/tcp/conn_pool.cc:364] [C18] connecting
Nov 25 15:42:45 ip-172-16-64-11 consul[198775]: [2020-11-25 15:42:45.300][198794][debug][connection] [external/envoy/source/common/network/connection_impl.cc:727] [C18] connecting to 192.168.42.209:20000
Nov 25 15:42:45 ip-172-16-64-11 consul[198775]: [2020-11-25 15:42:45.300][198794][debug][connection] [external/envoy/source/common/network/connection_impl.cc:736] [C18] connection in progress
Nov 25 15:42:45 ip-172-16-64-11 consul[198775]: [2020-11-25 15:42:45.300][198794][debug][pool] [external/envoy/source/common/tcp/conn_pool.cc:109] queueing request due to no available connections
Nov 25 15:42:45 ip-172-16-64-11 consul[198775]: [2020-11-25 15:42:45.300][198794][debug][conn_handler] [external/envoy/source/server/connection_handler_impl.cc:380] [C17] new connection
Nov 25 15:42:48 ip-172-16-64-11 consul[198775]: [2020-11-25 15:42:48.520][198775][debug][main] [external/envoy/source/server/server.cc:178] flushing stats
Nov 25 15:42:50 ip-172-16-64-11 consul[198775]: [2020-11-25 15:42:50.297][198794][debug][pool] [external/envoy/source/common/tcp/conn_pool.cc:403] [C18] connect timeout
Nov 25 15:42:50 ip-172-16-64-11 consul[198775]: [2020-11-25 15:42:50.297][198794][debug][connection] [external/envoy/source/common/network/connection_impl.cc:109] [C18] closing data_to_write=0 type=1
Nov 25 15:42:50 ip-172-16-64-11 consul[198775]: [2020-11-25 15:42:50.297][198794][debug][connection] [external/envoy/source/common/network/connection_impl.cc:200] [C18] closing socket: 1
Nov 25 15:42:50 ip-172-16-64-11 consul[198775]: [2020-11-25 15:42:50.297][198794][debug][pool] [external/envoy/source/common/tcp/conn_pool.cc:124] [C18] client disconnected
Nov 25 15:42:50 ip-172-16-64-11 consul[198775]: [2020-11-25 15:42:50.297][198794][debug][filter] [external/envoy/source/common/tcp_proxy/tcp_proxy.cc:535] [C17] connect timeout
Nov 25 15:42:50 ip-172-16-64-11 consul[198775]: [2020-11-25 15:42:50.297][198794][debug][filter] [external/envoy/source/common/tcp_proxy/tcp_proxy.cc:395] [C17] Creating connection to cluster grafana.default.dc3.internal.e5b63d86-f01b-3883-0b42-7ef5dd8700f9.consul
Nov 25 15:42:50 ip-172-16-64-11 consul[198775]: [2020-11-25 15:42:50.297][198794][debug][connection] [external/envoy/source/common/network/connection_impl.cc:109] [C17] closing data_to_write=0 type=1
Nov 25 15:42:50 ip-172-16-64-11 consul[198775]: [2020-11-25 15:42:50.297][198794][debug][connection] [external/envoy/source/common/network/connection_impl.cc:200] [C17] closing socket: 1
Nov 25 15:42:50 ip-172-16-64-11 consul[198775]: [2020-11-25 15:42:50.297][198794][debug][conn_handler] [external/envoy/source/server/connection_handler_impl.cc:86] [C17] adding to cleanup list
Nov 25 15:42:50 ip-172-16-64-11 consul[198775]: [2020-11-25 15:42:50.297][198794][debug][pool] [external/envoy/source/common/tcp/conn_pool.cc:238] [C18] connection destroyed
```

</details>

**Comment by rboyer** [comment]
> All the network connectivity is functionnal, each mesh-gateway can talk to any other one.

Can all of the servers in every datacenter directly reach all of the other servers in all other datacenters (just like the mesh gateways)? If kubernetes is in the mix typically this is extra tricky since the servers running in the k8s datacenter (dc3) will have PodIPs which may not be globally routable between your datacenters. In Consul 1.8.x there is a mechanism to work around this by having the servers _also_ talk to each other [via the mesh gateways](https://www.consul.io/docs/connect/gateways/mesh-gateway/wan-federation-via-mesh-gateways).

Additionally just deploying mesh gateways doesn't force service-to-service traffic to use them. That is controlled by a setting that can be set globally (via `proxy-defaults`) or on a service, or service instance basis. If you exclusively wanted service-to-service traffic that crosses between datacenters to go through the mesh gateways on both sides, then use this snippet:
```
Kind = "proxy-defaults",
Name = "global",
MeshGateway {
  mode = "local"
}
```

**Comment by rboyer** [comment]
For the server-to-server aspect, this very specific guide may help in your architecture: https://www.consul.io/docs/k8s/installation/multi-cluster/vms-and-kubernetes#kubernetes-as-the-secondary

**Comment by EhrhartMarius** [comment]
Hi,

>Can all of the servers in every datacenter directly reach all of the other servers in all other datacenters (just like the mesh gateways)?

No they cannot, only the mesh gateway is reachable. 

I know about the proxy-default and its applied both on my k8s via a value in the helm chart and on dc2 via the snipplet you just showed.

I also followed the tutorial you provided and all the configs are very similar from those shown in the documentation.

As for the upstream itself its very random, I got it to work sometimes but now it's fully stuck and besides resetting the whole mesh I don't have many options, and even then I don't know for sure that it would work.

I would be happy to provide more information.

**Comment by rboyer** [comment]
If you could include here the server configs for all 3 dcs, scrubbed of secrets that would help rule some stuff out.

Also run this from one representative server in all 3 dcs: `consul members -wan -detailed`

**Comment by mariusehr1** [comment]
<details>
  <summary>Client server dc2</summary>
```
data_dir = "/tmp/"
log_level = "TRACE"

datacenter = "dc2"
primary_datacenter = "aws"

server = true
bootstrap_expect = 1


bind_addr = "0.0.0.0"
client_addr = "0.0.0.0"
node_name = "consul-primary-dc2"


ports {
  grpc = 8502
  https = 8501
  http = -1
}
primary_gateways = ["mesh-gateway.domain.com:9443"]

connect {
    enabled = true
    enable_mesh_gateway_wan_federation = true
    ca_provider = "vault"
    ca_config {
        address = "https://vault.domain.com:8200"
        token = "s.token"
        root_pki_path = "connect-root"
        intermediate_pki_path = "dc2/connect-intermediate"
	ca_path = "/tls/aws-pub-vault.pem"
    }
}
acl {
  enabled = true
  default_policy = "deny"
  down_policy = "extend-cache"
  enable_token_persistence = true
  enable_token_replication = true
  tokens {
    replication="replication-token"
  }
}

encrypt = "encrypt key"

ui = true
enable_central_service_config = true

advertise_addr = "172.16.64.10"
verify_incoming = false
verify_outgoing= true
verify_server_hostname = true
ca_file = "/tls/aws-ca.pem"
cert_file = "/tls/aws-dc2.crt.pem"
key_file = "/tls/aws-dc2.key.pem"
auto_encrypt = {
    allow_tls = true
}
```
</details>

<details>
  <summary>Client server dc3 helm config </summary>
```
global:
  image: "consul:1.8.5"
  name: consul
  datacenter: dc3
  tls:
    enabled: true
    enableAutoEncrypt: true
    caCert:
      secretName: consul-ca-cert
      secretKey: tls.crt
    caKey:
      secretName: consul-ca-key
      secretKey: tls.key
  acls:
    manageSystemACLs: true
    bootstrapToken:
      secretName: consul-federation
      secretKey: replicationToken
    replicationToken:
      secretName: consul-federation
      secretKey: replicationToken
  federation:
    enabled: true
  gossipEncryption:
    secretName: consul-federation
    secretKey: gossipEncryptionKey
connectInject:
  enabled: true
  default: true
  envoyExtraArgs: "--log-level debug --disable-hot-restart"
  logLevel: "debug"
meshGateway:
  enabled: true
server:
  replicas: 2
  connect: true
  service:
    enabled : true
  bootstrapExpect: 2
  extraVolumes:
  - type: secret
    name: vault-config
    load: true
    items:
      - key: config
        path: vault-config.json
  - type: secret
    name: vault-ca
    load: false
  extraConfig: |
    {
      "primary_datacenter": "aws",
      "primary_gateways": ["mesh-gateway.domain.com:9443"]
    }
syncCatalog:
  enabled: true
  image: null
  default: true
  toConsul: true
  toK8S: true
  centralConfig:
    enabled: true
  aclSyncToken:
    secretName: consul-federation
    secretKey: replicationToken
```
</details>

<details>
  <summary>Client server primary hcl conf </summary>
```
domain = "consul."
primary_datacenter = "aws"

verify_incoming = true
verify_outgoing = true
verify_server_hostname = true
ca_path = "/opt/consul/tls/ca"
cert_file = "/opt/consul/tls/consul.crt.pem"
key_file = "/opt/consul/tls/consul.key.pem"

auto_encrypt {
    allow_tls = true
}

ports {
  grpc = 8502
  https = 8501
}

connect {
  enabled = true
  enable_mesh_gateway_wan_federation = true
}

enable_central_service_config = true
```
</details>

**Comment by rboyer** [comment]
A few more questions:

Can you check what all of the server TLS certs have as their [SAN fields] (https://www.consul.io/docs/connect/gateways/mesh-gateway/wan-federation-via-mesh-gateways#tls)?

They should all be something like :
```
server.<this_datacenter>.<domain>              (normal)
<node_name>.server.<this_datacenter>.<domain>  (needed for wan federation)
```

Also is the 4th datacenter still operational? I see a dc-sandbox in those membership listings.

**Comment by mariusehr1** [comment]
Yes, all the SAN fiels are properly setup for each datacenter. (They wouldn't federate if they didnt right? It would give an x509 error)

No, the 4th datacenter was just a test but I removed it.

**Comment by mariusehr1** [comment]
It seems that I am able to reach an other DC I created , from dc2 (composed of VMs as well, upstream is functionnal from both side).

I think that the problem is related to the Helm chart or kubernetes deployment in general. Which is really odd.

I am using the last version of the chart with the 1.8.5 version of consul.

I managed to get some logs from the envoy in the k8s trying to reach an other service in dc2 : 

```


[2020-11-27 17:24:53.642][11][debug][filter] [source/common/tcp_proxy/tcp_proxy.cc:251] [C95] new tcp proxy session
[2020-11-27 17:24:53.642][11][debug][filter] [source/common/tcp_proxy/tcp_proxy.cc:395] [C95] Creating connection to cluster local_app
[2020-11-27 17:24:53.642][11][debug][pool] [source/common/tcp/conn_pool.cc:83] creating a new connection
[2020-11-27 17:24:53.642][11][debug][pool] [source/common/tcp/conn_pool.cc:364] [C96] connecting
[2020-11-27 17:24:53.642][11][debug][connection] [source/common/network/connection_impl.cc:727] [C96] connecting to 127.0.0.1:0
[2020-11-27 17:24:53.642][11][debug][connection] [source/common/network/connection_impl.cc:736] [C96] connection in progress
[2020-11-27 17:24:53.642][11][debug][pool] [source/common/tcp/conn_pool.cc:109] queueing request due to no available connections
[2020-11-27 17:24:53.642][11][debug][conn_handler] [source/server/connection_handler_impl.cc:380] [C95] new connection
[2020-11-27 17:24:53.642][11][debug][connection] [source/extensions/transport_sockets/tls/ssl_socket.cc:198] [C95] handshake error: 5
[2020-11-27 17:24:53.642][11][debug][connection] [source/extensions/transport_sockets/tls/ssl_socket.cc:226] [C95] 
[2020-11-27 17:24:53.642][11][debug][connection] [source/common/network/connection_impl.cc:200] [C95] closing socket: 0
[2020-11-27 17:24:53.642][11][debug][pool] [source/common/tcp/conn_pool.cc:206] canceling pending request
[2020-11-27 17:24:53.642][11][debug][pool] [source/common/tcp/conn_pool.cc:214] canceling pending connection
[2020-11-27 17:24:53.642][11][debug][connection] [source/common/network/connection_impl.cc:109] [C96] closing data_to_write=0 type=1
[2020-11-27 17:24:53.642][11][debug][connection] [source/common/network/connection_impl.cc:200] [C96] closing socket: 1
[2020-11-27 17:24:53.642][11][debug][pool] [source/common/tcp/conn_pool.cc:124] [C96] client disconnected
[2020-11-27 17:24:53.642][11][debug][conn_handler] [source/server/connection_handler_impl.cc:86] [C95] adding to cleanup list
[2020-11-27 17:24:53.642][11][debug][pool] [source/common/tcp/conn_pool.cc:238] [C96] connection destroyed
```

Thanks for your help.

**Comment by mariusehr1** [comment]
I troubleshooted and found out about a couple things : 

- Any k8s DC will be able to reach services only within its own DC which leads me to believe its mesh-gateway related
- VM to VM datacenter communication is working fine on both sides.
- All the certificates are set with proper SAN.
- All the network has been checked as well as connectivity between the gateways and default mode is always local everywhere.
- I was using v0.25.0 of the Helm chart and updated it to v0.26.0 and changed a few image tag to get it to work, but the issue persisted, so I went back to v0.25.0.
- Every component is starting up as expected and no specific logs regarding permissions or encryption are shown.

The thing is I don't really know where to look and besides upstreams not working from/to k8s and a few logs in Envoy I can't really seem to find the reason why it doesn't work.
